### PR TITLE
fix(core)!: Allow syntax errors and expression errors to fail executions 

### PR DESCRIPTION
**Pull request by ivov**
https://linear.app/n8n/issue/N8N-6246

### CODE REVIEW

File: packages/workflow/src/Expression.ts [hunk 1]
```diff
@@ -22,20 +22,17 @@ import type { Workflow } from './Workflow';
...
 // Make sure that error get forwarded
 tmpl.tmpl.errorHandler = (error: Error) => {
-	if (error instanceof ExpressionError) {
-		if (error.context.failExecution) {
-			throw error;
-		}
-
-		if (typeof process === 'undefined' && error.clientOnly) {
-			throw error;
-		}
```
Comment (ivov) [inline packages/workflow/src/Expression.ts:38]: Not quite sure why we have this preexisting duplication and also whether `clientOnly` (used only once) is needed or can be simplified, but I'd rather keep this PR focused on the v1 changes.

File: packages/workflow/src/utils.ts [hunk 2]
```diff
@@ -127,3 +128,14 @@ export function assert<T>(condition: T, msg?: string): asserts condition {
 		throw error;
 	}
 }
+
+export const isFrontend = () => typeof process === 'undefined';
```
Comment (netroy) [inline packages/workflow/src/utils.ts:132]: do we need to evaluate this everytime? or do you think we could just assign this once?
```suggestion
export const isFrontend = typeof process === 'undefined';
```

File: packages/workflow/tsconfig.build.json [hunk 3]
```diff
@@ -7,6 +7,6 @@
 		"noEmit": false,
 		"tsBuildInfoFile": "dist/build.tsbuildinfo"
 	},
-	"include": ["src/**/*.ts"],
+	"include": ["src/**/*.ts", "package.json"],
```
Comment (netroy) [inline packages/workflow/tsconfig.build.json:10]: why is this change needed?

Comment (ivov) [reply packages/workflow/tsconfig.build.json:10]: But CI is [flagging this](https://github.com/n8n-io/n8n/actions/runs/5143041094/jobs/9257556793#step:6:31) anyway so the fix must be elsewhere.

**Review by netroy — DISMISSED** [review]
Can you please change the base branch of this PR to `release-v1`? Then you don't need to deal with these `package.json` issues.

File: packages/workflow/test/Expression.test.ts [hunk 4]
```diff
@@ -173,7 +173,18 @@ describe('Expression', () => {
...
+				value,
+				null,
+				0,
+				itemIndex,
+				'node',
+				data,
+				'manual',
+				'',
+				{},
+			);
```
Comment (ivov) [inline packages/workflow/test/Expression.test.ts:187]: After the change, [this test](https://github.com/n8n-io/n8n/blob/e8b51c8da9beecc1f4f95eb9528c8aa29f0f7222/packages/workflow/test/ExpressionFixtures/base.ts#L265-L269)

```ts
{
	type: 'evaluation',
	input: [],
	output: undefined,
},
```
started being caught by this check

```ts
if (executionData.length <= that.itemIndex) {
	throw new ExpressionError(`No data found for item-index: "${that.itemIndex}"`, {
		runIndex: that.runIndex,
		itemIndex: that.itemIndex,
	});
}
```

Because the itemIndex was hardcoded to `0`.

**Review by valya — APPROVED** [review]

**Review by netroy — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.0.5`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.0.5)
